File: test/test_functor.cpp
#include <ostream>
auto tuple_key_extractor = [](auto t) {return t + 1;};
struct tuple_t{
  unsigned int key;

  float value;

  tuple_t() 
  {
    this->key = 0;
    this->value = 0.0;
  }

  tuple_t(unsigned int key, float value) 
  {
    this->key = key;
    this->value = value;
  }

  friend std::ostream & operator<<(std::ostream & os, const tuple_t & d) 
  {
    os<<"key: "<<d.key<<","<<"value: "<<d.value<<"";
    return os;
  }

};
struct result_t{
  float sum;

  unsigned int count;

  float result;

  result_t() 
  {
    this->sum = 0.0;
    this->count = 0;
  }

  int   a() 
  {
    int a = 1;
    int bac[4] = {1, 2, 3, 4};
    return a;
  }

  float   mean(int a[10]) 
  {
    this->result = this->sum + this->count;
    return this->sum / this->count;
  }

  friend std::ostream & operator<<(std::ostream & os, const result_t & d) 
  {
    os<<"sum: "<<d.sum<<","<<"count: "<<d.count<<","<<"result: "<<d.result<<"";
    return os;
  }

};
struct window_functor{
  void   __call__(tuple_t tuple, result_t result) 
  {
    result.sum += tuple.value;
    result.count = result.count + 1;
  }

  tuple_t   test(tuple_t tuple) 
  {
    return tuple;
  }

};
int a();

int testl();

int fun(int a);

int a()
{
  int a = 1;
  return a;
}

int testl()
{
  auto z = [](auto x) {return x + 1;};
  int a = 1;
  auto t = tuple_key_extractor(1);
  return 1;
}

int fun(int a)
{
  int b = 0;
  int test = testl();
  return b;
}

File: test/test_functor.py
@transpile
class tuple_t:
    key : uint32  # maybe `uint` is better, but it is less C++ like
    value : float32     # or simply `float`

    # if not defined, the C++ default constructor is used
    def __init__(self):
        self.key = 0
        self.value = 0.0

    def __init__(self, key : uint32, value : float32):
        self.key = key
        self.value = value

tuple_key_extractor = lambda t: t+1

@transpile
class result_t:
    sum : float32    # or simply `float`
    count : uint32     # maybe `unsigned int` is better, but it is less C++ like
    # if not defined, the C++ default constructor is used
    def __init__(self):
        self.sum = 0.0
        self.count = 0

    # `float32` can be deduced from the context?
    def mean(self, a:[int,10]):
        def a(self):
            a = 1
            bac=[1,2,3,4]
            return a
        self.result=self.sum+self.count
        return self.sum / self.count

    # this is optional and it can be done after more important things are done
@transpile
class window_functor:
    i1:int
    c1:float32
    def __call__(self, tuple : tuple_t, result : result_t):
        result.sum += tuple.value
        result.count = result.count + 1
    def test(self,tuple:tuple_t):
        return tuple
@wireflow
def testl()->int:
    def a():
        a=1
        return a
    z=lambda x:x+1
    a=1
    t=tuple_key_extractor(1)
    return 1
@wireflow
def fun(a:int):
    b=0
    test=testl()
    return b
